src/UI/Actions.hs

```diff
@@ -289,9 +290,51 @@ instance HasList 'MailListOfAttachments where
 
 instance HasList 'ListOfFiles where
   type T 'ListOfFiles = Vector.Vector
-  type E 'ListOfFiles = (Bool, FileSystemEntry)
+  type E 'ListOfFiles = Toggleable FileSystemEntry
   list _ = asFileBrowser . fbEntries
 
+-- | contexts which have selectable items in a list
+class (HasList (n :: Name), Traversable (T n)) =>
+      HasToggleableList n
+  where
+  untoggleE :: Proxy n -> E n -> E n
+  toggleE :: Proxy n -> E n -> E n
+  isToggledE :: Proxy n -> E n -> Bool
+  untoggleAll :: (MonadState AppState m) => Proxy n -> m ()
+  untoggleAll proxy = modifying (list proxy . traversed) (untoggleE proxy)
+  toggle :: Proxy n -> Int -> StateT AppState (T.EventM Name) ()
+  -- | Traversal of selected items.  NOT a valid Traversal unless
+  -- selected state is preserved
+  toggledItemsL :: Proxy n -> Traversal' AppState (E n)
+  toggledItemsL proxy = list proxy . traversed . filtered (isToggledE proxy)
+  toggledOrSelectedItemHelper ::
```

Comment:
could we move this out of the class definition?

Reply:
Will do!